### PR TITLE
Add check for encryption in sdkconfig

### LINKED ISSUE #1067: The 'Flash device' button does not support encrypted mode (VSC-1213)
### OS

Windows

### Operating System version

Windows 10

### Visual Studio Code version

1.6.5

### ESP-IDF version

5.0.1

### Python version

3.11.1

### Doctor command output

---------------------------------------------- ESP-IDF Extension for Visual Studio Code report ---------------------------------------------
OS win32 x64 10.0.19045 
System environment variable IDF_PYTHON_ENV_PATH 
 undefined 
System environment variable PATH 
 C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Git\cmd;C:\Ruby31-x64\bin;C:\Program Files\PuTTY\;C:\ST\STM32CubeCLT;C:\ST\STM32CubeCLT\STM32CubeProgrammer\bin;C:\ST\STM32CubeCLT\STLink-gdb-server\bin;C:\ST\STM32CubeCLT\GNU-tools-for-STM32\bin;C:\Program Files (x86)\STMicroelectronics\stlink_server;C:\Program Files\Python311\Scripts\;C:\Program Files\Python311\;C:\Program Files\TortoiseGit\bin;C:<HOMEPATH>\AppData\Local\Microsoft\WindowsApps;C:<HOMEPATH>\AppData\Roaming\Python\Python311\Scripts 
System environment variable PYTHON 
 undefined 
Visual Studio Code version 1.83.1 
Visual Studio Code language en 
Visual Studio Code shell C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe 
ESP-IDF Extension version 1.6.5 
Workspace folder c:<HOMEPATH>\FTN\streetdog_ccu 
---------------------------------------------------- Extension configuration settings ------------------------------------------------------
ESP-ADF Path (idf.espAdfPath) ${env:ADF_PATH}
ESP-IDF Path (idf.espIdfPath) c:<HOMEPATH>\FTN\esp-idf-v5.0.1
ESP-MDF Path (idf.espMdfPath) ${env:MDF_PATH}
ESP-Matter Path (idf.espMatterPath) ${env:ESP_MATTER_PATH}
Custom extra paths (idf.customExtraPaths) C:<HOMEPATH>\.espressif\tools\xtensa-esp-elf-gdb\11.2_20220823\xtensa-esp-elf-gdb\bin;C:<HOMEPATH>\.espressif\tools\riscv32-esp-elf-gdb\11.2_20220823\riscv32-esp-elf-gdb\bin;C:<HOMEPATH>\.espressif\tools\xtensa-esp32-elf\esp-2022r1-11.2.0\xtensa-esp32-elf\bin;C:<HOMEPATH>\.espressif\tools\xtensa-esp32s2-elf\esp-2022r1-11.2.0\xtensa-esp32s2-elf\bin;C:<HOMEPATH>\.espressif\tools\xtensa-esp32s3-elf\esp-2022r1-11.2.0\xtensa-esp32s3-elf\bin;C:<HOMEPATH>\.espressif\tools\riscv32-esp-elf\esp-2022r1-11.2.0\riscv32-esp-elf\bin;C:<HOMEPATH>\.espressif\tools\esp32ulp-elf\2.35_20220830\esp32ulp-elf\bin;C:<HOMEPATH>\.espressif\tools\cmake\3.24.0\bin;C:<HOMEPATH>\.espressif\tools\openocd-esp32\v0.11.0-esp32-20221026\openocd-esp32\bin;C:<HOMEPATH>\.espressif\tools\ninja\1.10.2;C:<HOMEPATH>\.espressif\tools\idf-exe\1.0.3;C:<HOMEPATH>\.espressif\tools\ccache\4.6.2\ccache-4.6.2-windows-x86_64;C:<HOMEPATH>\.espressif\tools\dfu-util\0.9\dfu-util-0.9-win64;C:<HOMEPATH>\.espressif\tools\esp-rom-elfs\20220823
Custom extra vars (idf.customExtraVars)
    OPENOCD_SCRIPTS: C:<HOMEPATH>\.espressif\tools\openocd-esp32\v0.11.0-esp32-20221026/openocd-esp32/share/openocd/scripts
    IDF_CCACHE_ENABLE: 1
    ESP_ROM_ELF_DIR: C:<HOMEPATH>\.espressif\tools\esp-rom-elfs\20220823/
Virtual env Python Path (idf.pythonBinPath) C:<HOMEPATH>\.espressif\python_env\idf5.0_py3.8_env\Scripts\python.exe
Serial port (idf.port) COM5
OpenOCD Configs (idf.openOcdConfigs) interface/ftdi/esp32_devkitj_v1.cfg,board/esp32-wrover.cfg
ESP-IDF Tools Path (idf.toolsPath) C:<HOMEPATH>\.espressif
Git Path (idf.gitPath) C:<HOMEPATH>\.espressif\tools\idf-git\2.30.1\cmd\git.exe
-------------------------------------------------------- Configurations access -------------------------------------------------------------
Access to ESP-ADF Path (idf.espAdfPath) false
Access to ESP-IDF Path (idf.espIdfPath) true
Access to ESP-MDF Path (idf.espMdfPath) false
Access to ESP-Matter Path (idf.espMatterPath) false
Access to ESP-IDF Custom extra paths
Access to C:<HOMEPATH>\.espressif\tools\xtensa-esp-elf-gdb\11.2_20220823\xtensa-esp-elf-gdb\bin: true
Access to C:<HOMEPATH>\.espressif\tools\riscv32-esp-elf-gdb\11.2_20220823\riscv32-esp-elf-gdb\bin: true
Access to C:<HOMEPATH>\.espressif\tools\xtensa-esp32-elf\esp-2022r1-11.2.0\xtensa-esp32-elf\bin: true
Access to C:<HOMEPATH>\.espressif\tools\xtensa-esp32s2-elf\esp-2022r1-11.2.0\xtensa-esp32s2-elf\bin: true
Access to C:<HOMEPATH>\.espressif\tools\xtensa-esp32s3-elf\esp-2022r1-11.2.0\xtensa-esp32s3-elf\bin: true
Access to C:<HOMEPATH>\.espressif\tools\riscv32-esp-elf\esp-2022r1-11.2.0\riscv32-esp-elf\bin: true
Access to C:<HOMEPATH>\.espressif\tools\esp32ulp-elf\2.35_20220830\esp32ulp-elf\bin: true
Access to C:<HOMEPATH>\.espressif\tools\cmake\3.24.0\bin: true
Access to C:<HOMEPATH>\.espressif\tools\openocd-esp32\v0.11.0-esp32-20221026\openocd-esp32\bin: true
Access to C:<HOMEPATH>\.espressif\tools\ninja\1.10.2: true
Access to C:<HOMEPATH>\.espressif\tools\idf-exe\1.0.3: true
Access to C:<HOMEPATH>\.espressif\tools\ccache\4.6.2\ccache-4.6.2-windows-x86_64: true
Access to C:<HOMEPATH>\.espressif\tools\dfu-util\0.9\dfu-util-0.9-win64: true
Access to C:<HOMEPATH>\.espressif\tools\esp-rom-elfs\20220823: true
Access to Virtual env Python Path (idf.pythonBinPath) true
Access to CMake in environment PATH undefined
Access to Ninja in environment PATH undefined
Access to ESP-IDF Tools Path (idf.toolsPath) true
-------------------------------------------------------- Configurations has spaces -------------------------------------------------------------
Spaces in system environment Path true
Spaces in ESP-ADF Path (idf.espAdfPath) false
Spaces in ESP-IDF Path (idf.espIdfPath) false
Spaces in ESP-MDF Path (idf.espMdfPath) false
Spaces in ESP-Matter Path (idf.espMatterPath) false
Spaces in ESP-IDF Custom extra paths
Spaces in C:<HOMEPATH>\.espressif\tools\xtensa-esp-elf-gdb\11.2_20220823\xtensa-esp-elf-gdb\bin: false
Spaces in C:<HOMEPATH>\.espressif\tools\riscv32-esp-elf-gdb\11.2_20220823\riscv32-esp-elf-gdb\bin: false
Spaces in C:<HOMEPATH>\.espressif\tools\xtensa-esp32-elf\esp-2022r1-11.2.0\xtensa-esp32-elf\bin: false
Spaces in C:<HOMEPATH>\.espressif\tools\xtensa-esp32s2-elf\esp-2022r1-11.2.0\xtensa-esp32s2-elf\bin: false
Spaces in C:<HOMEPATH>\.espressif\tools\xtensa-esp32s3-elf\esp-2022r1-11.2.0\xtensa-esp32s3-elf\bin: false
Spaces in C:<HOMEPATH>\.espressif\tools\riscv32-esp-elf\esp-2022r1-11.2.0\riscv32-esp-elf\bin: false
Spaces in C:<HOMEPATH>\.espressif\tools\esp32ulp-elf\2.35_20220830\esp32ulp-elf\bin: false
Spaces in C:<HOMEPATH>\.espressif\tools\cmake\3.24.0\bin: false
Spaces in C:<HOMEPATH>\.espressif\tools\openocd-esp32\v0.11.0-esp32-20221026\openocd-esp32\bin: false
Spaces in C:<HOMEPATH>\.espressif\tools\ninja\1.10.2: false
Spaces in C:<HOMEPATH>\.espressif\tools\idf-exe\1.0.3: false
Spaces in C:<HOMEPATH>\.espressif\tools\ccache\4.6.2\ccache-4.6.2-windows-x86_64: false
Spaces in C:<HOMEPATH>\.espressif\tools\dfu-util\0.9\dfu-util-0.9-win64: false
Spaces in C:<HOMEPATH>\.espressif\tools\esp-rom-elfs\20220823: false
Spaces in Virtual env Python Path (idf.pythonBinPath) false
Spaces in ESP-IDF Tools Path (idf.toolsPath) false
----------------------------------------------------------- Executables Versions -----------------------------------------------------------
Git version 2.30.1.windows.1
ESP-IDF version 5.0.1
Python version 3.8.7
Python's pip version 23.1.2
-------------------------------------------------- Project configuration settings ----------------------------------------------------------
-------------------------------------------------- Python packages in idf.pythonBinPath ----------------------------------------------------
bitstring version: 4.0.1
CacheControl version: 0.12.11
certifi version: 2022.12.7
cffi version: 1.15.1
charset-normalizer version: 3.1.0
click version: 8.0.4
colorama version: 0.4.6
construct version: 2.10.68
contextlib2 version: 21.6.0
cryptography version: 36.0.2
ecdsa version: 0.18.0
esp-coredump version: 1.5.0
esptool version: 4.5.1
freertos-gdb version: 1.0.2
future version: 0.18.2
gcovr version: 6.0
idf-component-manager version: 1.2.3
idna version: 3.4
Jinja2 version: 3.0.3
kconfiglib version: 14.1.0
lockfile version: 0.12.2
lxml version: 4.9.2
MarkupSafe version: 2.1.2
msgpack version: 1.0.5
packaging version: 23.0
pip version: 23.1.2
psutil version: 5.9.5
pycparser version: 2.21
pyelftools version: 0.27
pygdbmi version: 0.9.0.2
Pygments version: 2.14.0
pyparsing version: 3.0.9
pyserial version: 3.5
pywin32 version: 306
PyYAML version: 6.0
reedsolo version: 1.6.0
requests version: 2.31.0
requests-file version: 1.5.1
requests-toolbelt version: 0.10.1
schema version: 0.7.5
setuptools version: 68.0.0
six version: 1.16.0
tqdm version: 4.65.0
urllib3 version: 1.26.15
websocket-client version: 1.6.0
wheel version: 0.36.2
windows-curses version: 2.3.1
xmlrunner version: 1.7.7
---------------------------------------------------- Check ESP-IDF python requirements.txt -------------------------------------------------
Check ESP-IDF Python packages c:<HOMEPATH>\FTN\esp-idf-v5.0.1\tools\check_python_dependencies.py:12: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
Python requirements are satisfied.
---------------------------------------------------- Check extension requirements.txt ------------------------------------------------------
Check Extension Python packages c:<HOMEPATH>\FTN\esp-idf-v5.0.1\tools\check_python_dependencies.py:12: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
Python requirements are satisfied.
---------------------------------------------------- Check ESP-IDF debug adapter requirements.txt ------------------------------------------
Check Debug AdapterPython packages Error
---------------------------------------------------- Visual Studio Code launch.json --------------------------------------------------------
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "espidf",
      "name": "Launch",
      "request": "launch"
    }
  ]
} 
---------------------------------------------------- Visual Studio Code c_cpp_properties.json ----------------------------------------------
{
    "configurations": [
        {
            "name": "Win32",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "defines": [
                "_DEBUG",
                "UNICODE",
                "_UNICODE"
            ],
            "compilerPath": "C:\\xxxxxx\\.espressif\\tools\\xtensa-esp32s3-elf\\esp-2022r1-11.2.0\\xtensa-esp32s3-elf\\bin\\xtensa-esp32s3-elf-gcc.exe"
        },
        {
            "name": "ESP-IDF",
            "cStandard": "c11",
            "cppStandard": "c++17",
            "includePath": [
                "C:\\xxxxxx\\esp-idf-v5.0.1/components/**",
                "c:\\xxxxxx\\yyyyyy/**"
            ],
            "browse": {
                "path": [
                    "C:\\xxxxxx\\esp-idf-v5.0.1/components",
                    "c:\\xxxxxx\\esp-idf-v5.0.1/components",
                    "${workspaceFolder}"
                ],
                "limitSymbolsToIncludedHeaders": true
            },
            "compilerPath": "C:\\xxxxxx\\.espressif\\tools\\xtensa-esp32s3-elf\\esp-2022r1-11.2.0\\xtensa-esp32s3-elf\\bin\\xtensa-esp32s3-elf-gcc.exe",
            "compileCommands": "c:\\xxxxxx\\streetdog_ccu/build/compile_commands.json",
            "compilerPathInCppPropertiesJson": "C:\\xxxxxx\\.espressif\\tools\\xtensa-esp32s3-elf\\esp-2022r1-11.2.0\\xtensa-esp32s3-elf\\bin\\xtensa-esp32s3-elf-gcc.exe",
            "compileCommandsInCppPropertiesJson": "${workspaceFolder}/build/compile_commands.json",
            "mergeConfigurations": false
        }
    ],
    "version": 4
} 
----------------------------------------------------------- Latest error -----------------------------------------------------------------
Latest error at Unknown error in ESP-IDF doctor command
--------------------------------------------------------------------------------------------------------------------------------------------

### Extension

_No response_

### Description

The project is configured for flash encryption (via menuconfig) and that is reflected in flasher_args.json but the command line does not set the 'encrypt' flag. 

flasher_args.json:

{
    "write_flash_args" : [ "--flash_mode", "dio",
                           "--flash_size", "16MB",
                           "--flash_freq", "80m" ],
    "flash_settings" : {
        "flash_mode": "dio",
        "flash_size": "16MB",
        "flash_freq": "80m"
    },
    "flash_files" : {
        "0x0" : "bootloader/bootloader.bin",
        "0x20000" : "app.bin",
        "0x8000" : "partition_table/partition-table.bin"
    },
    "bootloader" : { "offset" : "0x0", "file" : "bootloader/bootloader.bin", "encrypted" : "true" },
    "app" : { "offset" : "0x20000", "file" : "app.bin", "encrypted" : "true" },
    "partition-table" : { "offset" : "0x8000", "file" : "partition_table/partition-table.bin", "encrypted" : "true" },
    "extra_esptool_args" : {
        "after"  : "no_reset",
        "before" : "default_reset",
        "stub"   : true,
        "chip"   : "esp32s3"
    }
}

### Debug Message

```plain
*  Executing task: C:\xxxxx\.espressif\python_env\idf5.0_py3.8_env\Scripts\python.exe c:\xxxxxx\esp-idf-v5.0.1\components\esptool_py\esptool\esptool.py -p COM5 -b 460800 --before default_reset --after no_reset --chip esp32s3 write_flash --flash_mode dio --flash_freq 80m --flash_size 16MB 0x0 bootloader/bootloader.bin 0x20000 app.bin 0x8000 partition_table/partition-table.bin
```

### Other Steps to Reproduce

_No response_

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

### PR DESCRIPTION
**Pull request by radurentea**
## Description

Added check for SDK config modifications for encryption.

Fixes #1067 

## Type of change

Please delete options that are not relevant.
- Bug fix (non-breaking change which fixes an issue)

## Steps to test this pull request

1. Open a project
2. Use "ESP-IDF: SDK Configuration editor (menuconfig)
3. Enable "Enable flash encryption on boot (READ DOCS FIRST)"
4. Build and Flash project with UART
5. Flash should be successful and the task should contain the --encrypt flag
EX:

> Executing task: C:\xxxxx\.espressif\python_env\idf5.0_py3.8_env\Scripts\python.exe c:\xxxxxx\esp-idf-v5.0.1\components\esptool_py\esptool\esptool.py -p COM5 -b 460800 --before default_reset --after no_reset --chip esp32s3 write_flash --flash_mode dio --flash_freq 80m **--encrypt** --flash_size 16MB 0x0 bootloader/bootloader.bin 0x20000 app.bin 0x8000 partition_table/partition-table.bin

**Test Configuration**:
* ESP-IDF Version: 5.1
* OS (Windows,Linux and macOS): Mac

## Checklist
- [ ] PR Self Reviewed
- [ ] Applied Code formatting
- [ ] Added Documentation
- [ ] Added Unit Test
- [ ] Verified on all platforms - Windows,Linux and macOS

### CODE REVIEW

**Comment by owen63** [comment]
@radurentea @brianignacio5 How can I get this merged before it is closed (in 5 days)?

**Comment by kolipakakondal** [comment]
Hi @AndriiFilippov Please verify this

**Comment by AndriiFilippov** [comment]
@radurentea hi!

Tested: 
OS - Windows 10 
ESP-IDF: v5.1.1

error during build: 

```
[96/99] Linking C static library esp-idf\main\libmain.a
[97/99] Linking C executable bootloader.elf
[98/99] Generating binary image from built executable
esptool.py v4.7.dev3
Creating esp32 image...
Merged 1 ELF section
Successfully created esp32 image.
Generated C:/Users/AndriiFilippov/1/hello_world/build/bootloader/bootloader.bin
[99/99] cmd.exe /C "cd /D C:\Users\AndriiFilippov\1\hello_world\build\bootloader\esp-idf\esptool_py && c:\Users\AndriiFilippov\.espressif\python_env\idf5.1_py3.11_env\Scripts\python.exe C:/Users/AndriiFilippov/e/esp-idf-v5.1.1/components/partition_table/check_sizes.py --offset 0x8000 bootloader 0x1000 C:/Users/AndriiFilippov/1/hello_world/build/bootloader/bootloader.bin"
FAILED: esp-idf/esptool_py/CMakeFiles/bootloader_check_size C:/Users/AndriiFilippov/1/hello_world/build/bootloader/esp-idf/esptool_py/CMakeFiles/bootloader_check_size
cmd.exe /C "cd /D C:\Users\AndriiFilippov\1\hello_world\build\bootloader\esp-idf\esptool_py && c:\Users\AndriiFilippov\.espressif\python_env\idf5.1_py3.11_env\Scripts\python.exe C:/Users/AndriiFilippov/e/esp-idf-v5.1.1/components/partition_table/check_sizes.py --offset 0x8000 bootloader 0x1000 C:/Users/AndriiFilippov/1/hello_world/build/bootloader/bootloader.bin"
Error: Bootloader binary size 0x9350 bytes is too large for partition table offset 0x8000. Bootloader binary can be maximum 0x7000 (28672) bytes unless the partition table offset is increased in the Partition Table section of the project configuration menu.
ninja: build stopped: subcommand failed.
[189/879] Building C object esp-idf/esp_system/CMakeFiles/__idf_esp_system.dir/startup.c.obj
ninja: build stopped: subcommand failed.

 *  The terminal process "C:\WINDOWS\System32\WindowsPowerShell\v1.0\powershell.exe -Command ninja " terminated with exit code: 1. 
```

**Comment by radurentea** [comment]
Hi @AndriiFilippov,

It seems this is not a problem caused by VS Code and it is mentioned in the documentation link provided in the SDK Config Menu
image

Here is the part from the documentation link that is useful in this case:
> Enabling flash encryption will increase the size of bootloader, which might require updating partition table offset. See [Bootloader Size](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/bootloader.html#bootloader-size)

One of the solutions is to change the **Bootloader's log verbosity** from `Info`, which is default, to `Error`.

Screenshot 2023-11-22 143723

**Comment by AndriiFilippov** [comment]
@radurentea hi ! 

still error, but now related to encrypted flash write : 

```
Stub running...
Changing baud rate to 460800
Changed.
Configuring flash size...
Flash encryption key is not programmed

A fatal error occurred: Can't perform encrypted flash write, consult Flash Encryption documentation for more information

 *  The terminal process "c:\Users\AndriiFilippov\.espressif\python_env\idf5.1_py3.11_env\Scripts\python.exe 'c:\Users\AndriiFilippov\e\esp-idf-v5.1.1\components\esptool_py\esptool\esptool.py', '-p', 'COM9', '-b', '460800', '--before', 'default_reset', '--after', 'no_reset', '--chip', 'esp32s2', 'write_flash', '--flash_mode', 'dio', '--flash_freq', '80m', '--flash_size', '2MB', '--encrypt', '0x1000', 'bootloader/bootloader.bin', '0x10000', 'hello_world.bin', '0x8000', 'partition_table/partition-table.bin'" terminated with exit code: 2. 

```

**Comment by owen63** [comment]
My understanding is that when flash encryption is enabled (via menuconfig)
the esptool encrypt flag should be set, except when flashing onto a device
that doesn't have the encrypt fuse set. In this case the bootloader detects
it and dynamically encrypts the firmware at first boot and sets the fuse.
So the 'Flash device' button must first query the target encrypt fuse and
set the esptool flag accordingly.

File: src/extension.ts [hunk 1]
```diff
@@ -3251,8 +3251,17 @@ const build = (flashType?: ESP.FlashType) => {
     );
   });
 };
+
+function isFlashEncryptionEnabled() {
```
Comment (brianignacio5) [inline src/extension.ts:3255]: Maybe we can move this function to flash related files ?

File: src/extension.ts [hunk 2]
```diff
@@ -3389,6 +3401,33 @@ async function startFlashing(
     flashType = await selectFlashMethod();
   }
 
+  Logger.info(`Using flash type: ${flashType}`, {tag: "Flash"});
```
Comment (brianignacio5) [inline src/extension.ts:3404]: Also this we could move it to the flashing related files.

File: src/extension.ts [hunk 3]
```diff
@@ -3389,6 +3401,33 @@ async function startFlashing(
...
+
+  try {
+    if(encryptPartitions) {
+      if(flashType !== ESP.FlashType.UART) {
+        throw new Error(`Invalid flash type for partition encryption. Required: UART, Found: ${flashType}`);
+      }
+      const eFuse = new ESPEFuseManager(workspaceRoot);
+      const summaryResult = await eFuse.readSummary();
+      if (summaryResult && summaryResult.BLOCK_KEY0) {
+        if (summaryResult.BLOCK_KEY0.value && summaryResult.BLOCK_KEY0.value == "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00") {
```
Comment (brianignacio5) [inline src/extension.ts:3414]: I would suggest to move the BLOCK_KEY0 default value to config.ts as a global variable and provide a meaningful name like empty eFuse or blank or something.

File: src/utils.ts [hunk 4]
```diff
@@ -1198,3 +1198,24 @@ export function markdownToWebviewHtml(
...
 }
+
+/**
+ * Shows an error notification with a button that opens a link when clicked.
+ * @param {string} errorMessage - The error message to display.
+ * @param {string} [buttonLabel="Read Documentation"] - The label for the button (default: "Read Documentation")
+ * @param {string} linkUrl - The URL to open when the button is clicked.
+ * @returns {Promise<void>} - A promise that resolves when the notification is shown.
+ */
+export async function showErrorNotificationWithLink(
```
Comment (brianignacio5) [inline src/utils.ts:1209]: This function could live closer to the notification logic like `src/logger` and extended for other use cases.

**Review by brianignacio5 — CHANGES_REQUESTED** [review]
A few changes to reduce code and refactor code closer to functionality separation.

**Comment by AndriiFilippov** [comment]
@radurentea hi !

Tested under: 

OS - Windows 10 
ESP-IDF: v5.1.2 

Still getting error during flash: 


```
Configuring flash size...
Unexpected FLASH_CRYPT_CONFIG value: 0x0
Flash encryption key is not programmed

A fatal error occurred: Can't perform encrypted flash write, consult Flash Encryption documentation for more information

 *  The terminal process "c:\Users\AndriiFilippov\.espressif\python_env\idf5.1_py3.11_env\Scripts\python.exe 'c:\Users\AndriiFilippov\e\esp-idf-v5.1.2\components\esptool_py\esptool\esptool.py', '-p', 'COM13', '-b', '460800', '--before', 'default_reset', '--after', 'no_reset', '--chip', 'esp32', 'write_flash', '--flash_mode', 'dio', '--flash_freq', '40m', '--flash_size', '2MB', '--encrypt', '0x1000', 'bootloader/bootloader.bin', '0x20000', 'project-nameOUY.bin', '0x9000', 'partition_table/partition-table.bin'" terminated with exit code: 2. 
```

File: src/flash/flashCmd.ts [hunk 5]
```diff
@@ -88,33 +104,128 @@ export async function verifyCanFlash(
...
+        resultType: FlashCheckResultType.ErrorInvalidFlashType,
+      };
+    }
+    const eFuse = new ESPEFuseManager(workspaceRoot);
+    const summaryResult = await eFuse.readSummary();
+    if (summaryResult && summaryResult.BLOCK_KEY0) {
+      if (
+        // eFuse is not set
+        summaryResult.BLOCK_KEY0.value &&
+        summaryResult.BLOCK_KEY0.value === ESP.DEFAULT_UNSET_EFUSE_VALUE
```
Comment (brianignacio5) [inline src/flash/flashCmd.ts:199]: The reference block you use here is not available for esp32 on ESP-IDF v5.0.4.

In the eFuse summary we receive:

```ts
BLOCK1: {
  bit_len: 256,
  block: 1,
  category: "security",
  description: "Flash encryption key",
  efuse_type: "bytes:32",
  name: "BLOCK1",
  pos: 0,
  readable: true,
  value: "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00",
  word: 0,
  writeable: true,
}
```

This causes the flash to execute as encrypted and fail because of missing eFuse:

```
Executing task: /Users/brainignacio/.espressif/python_env/idf5.0_py3.11_env/bin/python /Users/brainignacio/esp/esp504/esp-idf/components/esptool_py/esptool/esptool.py -p /dev/cu.usbserial-0001 -b 460800 --before default_reset --after no_reset --chip esp32 write_flash --flash_mode dio --flash_freq 40m --flash_size 2MB --encrypt 0x1000 bootloader/bootloader.bin 0x20000 flash_encryption.bin 0x10000 partition_table/partition-table.bin 

esptool.py v4.7.0
Serial port /dev/cu.usbserial-0001
Connecting......
Chip is ESP32-D0WD (revision v0.0)
Features: WiFi, BT, Dual Core, Coding Scheme None
Crystal is 40MHz
MAC: 30:ae:a4:00:45:cc
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
Configuring flash size...
Unexpected FLASH_CRYPT_CONFIG value: 0x0
Flash encryption key is not programmed

A fatal error occurred: Can't perform encrypted flash write, consult Flash Encryption documentation for more information
```

File: src/flash/flashCmd.ts [hunk 6]
```diff
@@ -88,33 +104,128 @@ export async function verifyCanFlash(
...
+        showInfoNotificationWithLink(warnMessage, documentationUrl);
+        OutputChannel.appendLineAndShow(warnMessage, "Flash Encryption");
+        Logger.info(warnMessage, { tag: "Flash Encryption" });
+        return {
+          success: false,
+          resultType: FlashCheckResultType.ErrorEfuseNotSet,
+        };
+      }
+    }
+    return { success: true };
```
Comment (brianignacio5) [inline src/flash/flashCmd.ts:217]: This way the encrypt flashing will be executed even if the eFuse is not found, I think we should make sure that encryption is enabled in order to avoid accident encryption flashing

File: src/espIdf/unitTest/adapter.ts [hunk 7]
```diff
@@ -146,8 +146,7 @@ export class UnitTest {
       TestRunProfileKind.Run,
       runHandler,
       true,
-      undefined,
-      false
+      undefined
```
Comment (brianignacio5) [inline src/espIdf/unitTest/adapter.ts:149]: Why is this file modified in this PR ?

File: src/espIdf/documentation/getDocsVersion.ts [hunk 8]
```diff
@@ -111,3 +113,36 @@ export async function readObjectFromUrlFile(objectUrl: string) {
...
+    "idf.espIdfPath",
+    workspace
+  ) as string;
+
+  const adapterTargetName =
+    (idfConf.readParameter("idf.adapterTargetName", workspace) as string) ||
+    "esp32";
+  const idfVersion = await getEspIdfFromCMake(espIdfPath);
+  const docVersions = await getDocsVersion();
+  let docVersion = docVersions.find((docVer) => docVer.name === idfVersion);
```
Comment (brianignacio5) [inline src/espIdf/documentation/getDocsVersion.ts:137]: Will suggest to add validation in case idfVersion doesn't exist.

**Review by brianignacio5 — CHANGES_REQUESTED** [review]
Obtained same error due to unset flash encryption eFuse, please review.

File: src/flash/flashCmd.ts (hunk 5 above)
Comment (radurentea) [reply src/flash/flashCmd.ts]: I've updated this part to search for the property:  "FLASH_CRYPT_CNT" after running espefuse summary command and "SPI_BOOT_CRYPT_CNT" for all the other boards

File: src/espIdf/documentation/getDocsVersion.ts (hunk 8 above)
Comment (radurentea) [reply src/espIdf/documentation/getDocsVersion.ts:137]: I've added validation for a better user experience. User is warned if idfVersion is not found, the link for the documentation will contain the latest version

File: src/flash/flashCmd.ts (hunk 6 above)
Comment (radurentea) [reply src/flash/flashCmd.ts]: Good catch! Fixed it.

File: src/espIdf/unitTest/adapter.ts (hunk 7 above)
Comment (radurentea) [reply src/espIdf/unitTest/adapter.ts]: Reverted this change as it was indeed unnessecary.

**Comment by radurentea** [comment]
Closing this PR because I've re-implemented this here: https://github.com/espressif/vscode-esp-idf-extension/pull/1225

The reason is that the branch is too old and there are too many conflicts to solve.

**Comment by brianignacio5** [comment]
> Closing this PR because I've re-implemented this here: #1225
> 
> The reason is that the branch is too old and there are too many conflicts to solve.

The linked PR #1225  is not related to this PR I believe @radurentea . Can we retake work here in a new PR ?

**Comment by radurentea** [comment]
Hi @brianignacio5, I mean to put this message in the hints viewer's first implementation.

Yes, I will create a new PR for this one as well.
